## 1)

In [67]:
%sql select * from pg_indexes where tablename = 'circuits';

3 rows affected.


[('f1db', 'circuits', 'idx_17102_url', None, 'CREATE UNIQUE INDEX idx_17102_url ON circuits USING btree (url)'),
 ('f1db', 'circuits', 'circuits_position_idx', None, 'CREATE INDEX circuits_position_idx ON circuits USING gist ("position")'),
 ('f1db', 'circuits', 'idx_17102_primary', None, 'CREATE UNIQUE INDEX idx_17102_primary ON circuits USING btree (circuitid)')]

There are three indices on the 'circuits' table. There are two default btress and a gist index which makes sense since it is to do with positioning.

In the appdev, writing '\di+' + the indexname, will show us the size of the indices.

| indexname             | size       |
|-----------------------|------------|
| idx_17102_url         | 16 bytes   |
| circuits_position_idx | 8192 bytes |
| idx_17102_primary     | 16 bytes   |

## 2)

In [63]:
%%sql SELECT races.name AS races_name, races.date, results.milliseconds, drivers.surname, EXTRACT(YEAR from AGE(CURRENT_DATE, drivers.dob)) AS age, circuits.name AS circuit_name FROM races
JOIN results USING (raceid)
JOIN drivers USING (driverid)
JOIN circuits USING (circuitid)
WHERE EXTRACT(YEAR from AGE(CURRENT_DATE, drivers.dob)) < 38
AND results.position = 1 
ORDER BY drivers.surname ASC;

184 rows affected.


[('German Grand Prix', datetime.date(2010, 7, 25), 5258864, 'Alonso', 36.0, 'Hockenheimring'),
 ('Bahrain Grand Prix', datetime.date(2010, 3, 14), 5960396, 'Alonso', 36.0, 'Bahrain International Circuit'),
 ('Malaysian Grand Prix', datetime.date(2005, 3, 20), 5493736, 'Alonso', 36.0, 'Sepang International Circuit'),
 ('Bahrain Grand Prix', datetime.date(2005, 4, 3), 5358531, 'Alonso', 36.0, 'Bahrain International Circuit'),
 ('San Marino Grand Prix', datetime.date(2005, 4, 24), 5261921, 'Alonso', 36.0, 'Autodromo Enzo e Dino Ferrari'),
 ('European Grand Prix', datetime.date(2005, 5, 29), 5506648, 'Alonso', 36.0, 'Nürburgring'),
 ('French Grand Prix', datetime.date(2005, 7, 3), 5482232, 'Alonso', 36.0, 'Circuit de Nevers Magny-Cours'),
 ('German Grand Prix', datetime.date(2005, 7, 24), 5188599, 'Alonso', 36.0, 'Hockenheimring'),
 ('Chinese Grand Prix', datetime.date(2005, 10, 16), 5993618, 'Alonso', 36.0, 'Shanghai International Circuit'),
 ('Hungarian Grand Prix', datetime.date(2003, 8, 24), 5941460, 'Alonso', 36.0, 'Hungaroring'),
 ('Chinese Grand Prix', datetime.date(2013, 4, 14), 5786945, 'Alonso', 36.0, 'Shanghai International Circuit'),
 ('Singapore Grand Prix', datetime.date(2008, 9, 28), 7036304, 'Alonso', 36.0, 'Marina Bay Street Circuit'),
 ('Japanese Grand Prix', datetime.date(2008, 10, 12), 5421892, 'Alonso', 36.0, 'Fuji Speedway'),
 ('Spanish Grand Prix', datetime.date(2013, 5, 12), 5956596, 'Alonso', 36.0, 'Circuit de Barcelona-Catalunya'),
 ('Canadian Grand Prix', datetime.date(2006, 6, 25), 5677308, 'Alonso', 36.0, 'Circuit Gilles Villeneuve'),
 ('Japanese Grand Prix', datetime.date(2006, 10, 8), 5032413, 'Alonso', 36.0, 'Suzuka Circuit'),
 ('Malaysian Grand Prix', datetime.date(2007, 4, 8), 5534930, 'Alonso', 36.0, 'Sepang International Circuit'),
 ('Korean Grand Prix', datetime.date(2010, 10, 24), 10100810, 'Alonso', 36.0, 'Korean International Circuit'),
 ('Singapore Grand Prix', datetime.date(2010, 9, 26), 7073579, 'Alonso', 36.0, 'Marina Bay Street Circuit'),
 ('Monaco Grand Prix', datetime.date(2007, 5, 27), 6029329, 'Alonso', 36.0, 'Circuit de Monaco'),
 ('German Grand Prix', datetime.date(2012, 7, 22), 5465862, 'Alonso', 36.0, 'Hockenheimring'),
 ('European Grand Prix', datetime.date(2012, 6, 24), 6256649, 'Alonso', 36.0, 'Valencia Street Circuit'),
 ('European Grand Prix', datetime.date(2007, 7, 22), 7586358, 'Alonso', 36.0, 'Nürburgring'),
 ('Malaysian Grand Prix', datetime.date(2012, 3, 25), 9891812, 'Alonso', 36.0, 'Sepang International Circuit'),
 ('Italian Grand Prix', datetime.date(2010, 9, 12), 4584572, 'Alonso', 36.0, 'Autodromo Nazionale di Monza'),
 ('Italian Grand Prix', datetime.date(2007, 9, 9), 4717806, 'Alonso', 36.0, 'Autodromo Nazionale di Monza'),
 ('British Grand Prix', datetime.date(2011, 7, 10), 5321196, 'Alonso', 36.0, 'Silverstone Circuit'),
 ('Bahrain Grand Prix', datetime.date(2006, 3, 12), 5386205, 'Alonso', 36.0, 'Bahrain International Circuit'),
 ('Australian Grand Prix', datetime.date(2006, 4, 2), 5667870, 'Alonso', 36.0, 'Albert Park Grand Prix Circuit'),
 ('Spanish Grand Prix', datetime.date(2006, 5, 14), 5181759, 'Alonso', 36.0, 'Circuit de Barcelona-Catalunya'),
 ('Monaco Grand Prix', datetime.date(2006, 5, 28), 6223116, 'Alonso', 36.0, 'Circuit de Monaco'),
 ('British Grand Prix', datetime.date(2006, 6, 11), 5151927, 'Alonso', 36.0, 'Silverstone Circuit'),
 ('Russian Grand Prix', datetime.date(2017, 4, 30), 5288743, 'Bottas', 28.0, 'Sochi Autodrom'),
 ('Austrian Grand Prix', datetime.date(2017, 7, 9), 4908523, 'Bottas', 28.0, 'Red Bull Ring'),
 ('Bahrain Grand Prix', datetime.date(2014, 4, 6), 5982743, 'Hamilton', 33.0, 'Bahrain International Circuit'),
 ('Spanish Grand Prix', datetime.date(2014, 5, 11), 6065155, 'Hamilton', 33.0, 'Circuit de Barcelona-Catalunya'),
 ('British Grand Prix', datetime.date(2008, 7, 6), 5949440, 'Hamilton', 33.0, 'Silverstone Circuit'),
 ('British Grand Prix', datetime.date(2014, 7, 6), 8812094, 'Hamilton', 33.0, 'Silverstone Circuit'),
 ('G

## 3)

In [66]:
%%sql EXPLAIN ANALYZE SELECT races.name AS races_name, races.date, results.milliseconds, drivers.surname, EXTRACT(YEAR from AGE(CURRENT_DATE, drivers.dob)) AS age, circuits.name AS circuit_name FROM races
JOIN results USING (raceid)
JOIN drivers USING (driverid)
JOIN circuits USING (circuitid)
WHERE EXTRACT(YEAR from AGE(CURRENT_DATE, drivers.dob)) < 38
AND results.position = 1 
ORDER BY drivers.surname ASC;

25 rows affected.


[('Sort  (cost=829.08..829.89 rows=324 width=66) (actual time=24.540..25.131 rows=184 loops=1)',),
 ('  Sort Key: drivers.surname',),
 ('  Sort Method: quicksort  Memory: 50kB',),
 ('  ->  Hash Join  (cost=82.13..815.57 rows=324 width=66) (actual time=8.304..23.798 rows=184 loops=1)',),
 ('        Hash Cond: (races.circuitid = circuits.circuitid)',),
 ('        ->  Hash Join  (cost=78.48..803.83 rows=324 width=50) (actual time=7.640..21.680 rows=184 loops=1)',),
 ('              Hash Cond: (results.raceid = races.raceid)',),
 ('              ->  Hash Join  (cost=38.53..759.78 rows=324 width=27) (actual time=1.137..13.835 rows=184 loops=1)',),
 ('                    Hash Cond: (results.driverid = drivers.driverid)',),
 ('                    ->  Seq Scan on results  (cost=0.00..708.96 rows=974 width=24) (actual time=0.010..8.446 rows=974 loops=1)',),
 ('                          Filter: ("position" = 1)',),
 ('                          Rows Removed by Filter: 22703',),
 ('                    ->  Hash  (cost=35.03..35.03 rows=280 width=19) (actual time=1.107..1.107 rows=60 loops=1)',),
 ('                          Buckets: 1024  Batches: 1  Memory Usage: 12kB',),
 ('                          ->  Seq Scan on drivers  (cost=0.00..35.03 rows=280 width=19) (actual time=0.015..0.710 rows=60 loops=1)',),
 ("                                Filter: (date_part('year'::text, age((CURRENT_DATE)::timestamp with time zone, (dob)::timestamp with time zone)) < '38'::double precision)",),
 ('                                Rows Removed by Filter: 781',),
 ('              ->  Hash  (cost=27.76..27.76 rows=976 width=39) (actual time=6.482..6.482 rows=976 loops=1)',),
 ('                    Buckets: 1024  Batches: 1  Memory Usage: 79kB',),
 ('                    ->  Seq Scan on races  (cost=0.00..27.76 rows=976 width=39) (actual time=0.010..3.238 rows=976 loops=1)',),
 ('        ->  Hash  (cost=2.73..2.73 rows=73 width=28) (actual time=0.643..0.643 rows=73 loops=1)',),
 ('              Buckets: 1024  Batches: 1  Memory Usage: 13kB',),
 ('              ->  Seq Scan on circuits  (cost=0.00..2.73 rows=73 width=28) (actual time=0.012..0.311 rows=73 loops=1)',),
 ('Planning time: 0.438 ms',),
 ('Execution time: 25.787 ms',)]

I am unsure, but I assume that the calls are the rows that has as -> in front of the line.
In that case, we are making 10 calls. First the system is sorting by drivers.surname, then the system does the joining of the tables in question, ending up with the removal of 22703 rows. Then additional rows are removed by the filtering from our query, until we have the required rows that were requested.
The planning time was 0.438 ms and the execution time was 25.787 ms.

## 4)

In [73]:
%%sql CREATE MATERIALIZED VIEW bra_view_cache AS
SELECT 
races.name AS races_name, races.date, results.milliseconds, drivers.surname, EXTRACT(YEAR from AGE(CURRENT_DATE, drivers.dob)) AS age, circuits.name AS circuit_name FROM races
JOIN results USING (raceid)
JOIN drivers USING (driverid)
JOIN circuits USING (circuitid)
WHERE EXTRACT(YEAR from AGE(CURRENT_DATE, drivers.dob)) < 38
AND results.position = 1 
ORDER BY drivers.surname ASC;

184 rows affected.


[]

In [74]:
%sql EXPLAIN ANALYZE SELECT * FROM bra_view_cache;

3 rows affected.


[('Seq Scan on bra_view_cache  (cost=0.00..10.50 rows=50 width=1568) (actual time=0.013..0.792 rows=184 loops=1)',),
 ('Planning time: 0.135 ms',),
 ('Execution time: 1.713 ms',)]

After having created the materialized view, we now have a cached version of our query results. This means that we can access the results a lot faster, since the results are now cached. A downside to this could be that results that for some reason should change, will not be included in the cached results. 'REFRESH MATERIALIZED VIEW' should be used to update these results whenever deemed necessary.
Obviously cached results will affect the execution time, as can be seen from the explain-analyze query above. The planning time is now 0.135 ms and the execution time is 1.713 ms.